### PR TITLE
Added version 4.3.0 of libcoap.

### PR DESCRIPTION
**Pull request by samvik**
Specify library name and version:  **libcoap/4.3.0**

The only currently existing version of libcoap is not from an official release and more than a year old. For the project I'm working on we need a newer version.

~~I have created the recipe for 4.3.0 in a separate directory since both the library name and headers have changed in this release. Is this the correct way of doing it or is it better to have one recipe with some conditional statements?~~
As per review comment, the existing recipe is updated.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
Thanks for your contribution, we only change folders when we need to break the recipe. You still should reuse the current recipe for both versions, don't need to create a new folder.

File: recipes/libcoap/all/conanfile.py [hunk 1]
```diff
@@ -60,9 +60,14 @@ def configure(self):
 
     def source(self):
         tools.get(**self.conan_data["sources"][self.version])
```
Comment (uilianries) [inline recipes/libcoap/all/conanfile.py:62]:
```suggestion
        tools.get(**self.conan_data["sources"][self.version],
                  destination=self._source_subfolder, strip_root=True)
```

You can replace all source(self) content using it. The destination + strip_root, will do all magic related to folder name.

**Review by uilianries — COMMENTED** [review]
Almost good to go.

File: recipes/libcoap/all/conanfile.py (hunk 1 above)
Comment (samvik) [reply recipes/libcoap/all/conanfile.py:62]: Sweet!

**Review by uilianries — APPROVED** [review]
Thank you!

File: recipes/libcoap/all/conanfile.py [hunk 2]
```diff
@@ -86,12 +88,21 @@ def package(self):
...
+        else:
+            library_name = "coap-3"
+            pkgconfig_name = "libcoap-3"
+
         self.cpp_info.components["coap"].names["cmake_find_package"] = "coap"
         self.cpp_info.components["coap"].names["cmake_find_package_multi"] = "coap"
-        self.cpp_info.components["coap"].names["pkg_config"] = pkgconfig_filename
-        self.cpp_info.components["coap"].libs = ["coap"]
+        pkgconfig_filename = "{}{}".format(pkgconfig_name, "-{}".format(self.options.dtls_backend) if self.options.dtls_backend else "")
+        self.cpp_info.names["pkg_config"] = pkgconfig_filename
```
Comment (samvik) [inline recipes/libcoap/all/conanfile.py:103]: It looks like I have managed to remove one line setting the pkgconfig_file. In the original recipe both these lines where present. Did not realize there where two places this information where set.

self.cpp_info.names["pkg_config"] = pkgconfig_filename 
self.cpp_info.components["coap"].names["pkg_config"] = pkgconfig_filename

I assume I should re-add the missing one, please advice.

Comment (prince-chrismc) [reply recipes/libcoap/all/conanfile.py:103]: You should only need one and the component one should be used :) 

Good catch!

Comment (samvik) [reply recipes/libcoap/all/conanfile.py]: It's now fixed.

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**